identity/src/main/java/com/android/identity/util/SimpleApplicationData.java

```diff
@@ -0,0 +1,257 @@
+        }
+        mListener = listener;
+        mListenerExecutor = executor;
+    }
+
+    @Override
+    public ApplicationData setData(@NonNull String key, @Nullable byte[] value) {
+        if (value == null) {
+            mApplicationData.remove(key);
+        } else {
+            mApplicationData.put(key, value);
+        }
+
+        if (mListener != null && mListenerExecutor != null) {
+            mListenerExecutor.execute(mListener::onDataSet);
+        }
+        return this;
+    }
+
+    @Override
+    public ApplicationData setString(@NonNull String key, @NonNull String value) {
+        if (value == null) {
+            throw new IllegalArgumentException("Value cannot be null");
+        } else {
+            mApplicationData.put(key, Util.cborEncodeString(value));
+        }
+
+        if (mListener != null && mListenerExecutor != null) {
+            mListenerExecutor.execute(mListener::onDataSet);
+        }
+        return this;
+    }
+
+    @Override
+    public ApplicationData setNumber(@NonNull String key, long value) {
+        mApplicationData.put(key, Util.cborEncodeNumber(value));
+
+        if (mListener != null && mListenerExecutor != null) {
+            mListenerExecutor.execute(mListener::onDataSet);
+        }
+        return this;
+    }
+
+    @Override
+    public ApplicationData setTimestamp(@NonNull String key, @NonNull Timestamp value) {
+        if (value == null) {
+            throw new IllegalArgumentException("Value cannot be null");
+        } else {
+            mApplicationData.put(key, Util.cborEncodeNumber(value.toEpochMilli()));
+        }
+
+        if (mListener != null && mListenerExecutor != null) {
+            mListenerExecutor.execute(mListener::onDataSet);
+        }
+        return this;
+    }
+
+    @Override
+    public ApplicationData setBoolean(@NonNull String key, boolean value) {
+        mApplicationData.put(key, Util.cborEncodeBoolean(value));
+
+        if (mListener != null && mListenerExecutor != null) {
+            mListenerExecutor.execute(mListener::onDataSet);
+        }
+        return this;
+    }
+
+    @Override
+    public boolean keyExists(@NonNull String key) {
+        return mApplicationData.get(key) != null;
+    }
+
+    @NonNull
+    @Override
+    public byte[] getData(@NonNull String key) {
+        byte[] value = mApplicationData.get(key);
+        if (value == null) {
+            throw new IllegalArgumentException("This key is not present in the ApplicationData.");
+        }
+        return value;
+    }
+
+    @NonNull
+    @Override
+    public String getString(@NonNull String key) {
+        byte[] value = mApplicationData.get(key);
+        if (value == null) {
+            throw new IllegalArgumentException("This key is not present in the ApplicationData.");
+        }
+        if (value.length == 0) {
+            return "";
+        }
+        return Util.cborDecodeString(value);
+    }
```

Comment:
I appreciate the call-out to avoid repeated checks!

For the check `value.length == 0`, I added it to ensure the expected behavior in `com/android/identity/credential/CredentialStoreTest.java` lines up with actual behavior. For reference, the lines that indicate this expected behavior are pasted below:
```
pendingAppData.setString("foo", value);
pendingAppData.setData("bar", new byte[0]);
Assert.assertEquals(value, pendingAppData.getString("foo"));
Assert.assertEquals(0, pendingAppData.getString("bar").length());
```
Without the explicit check, `Util.cborDecodeString` fails with `Unexpected number of items, expected 1 got 0`

I'm personally not committed to the expected behavior of appData.getString on an empty byte array to return an empty String, but ik that's what you expected when you wrote the tests. I can either change the tests in `com/android/identity/credential/CredentialStoreTest.java` or leave the check for `value.length == 0` - up to you!